### ISSUE TITLE
Common reward model interface

### ISSUE THREAD

**Issue by AdamGleave** [HumanCompatibleAI/imitation#246]
Currently our `RewardNet` class is designed for `AIRL`. In particular, having separate `train` and `test` outputs is specific to AIRL, and does not make any sense for most others algorithms.

We currently have an alternative interface `tabular_irl.RewardModel`. This is intended for JAX -- but it'd be incompatible with `RewardNet` even after being ported to PyTorch. (And note that it's not actually specific to tabular methods -- although some of the concrete subclasses are.)

In the future we might add other algorithms like GCL, DRLHP, etc which also use rewards but do not have this train vs test distinction.

My proposal would be to make an abstract `RewardNet` or `RewardModel` class (either name seems fine) that is minimal and so should work across all algorithms. In particular, I imagine the `__init__` would take in an `observation_space` and `action_space`, there would be a single abstract method `reward` (v.s. `reward_train` and `reward_test` now) and a concrete implementation `predict` taking NumPy arrays. (You might also need some other things for PyTorch, e.g. `device`, that we have now.)

We can then have concrete instances for various different reward models. I would suggest having one that is an MLP reward model (with boolean flags indicating whether it uses observation, action, next observation and dones as input), one that is an MLP potential shaping only, a class that lets you take linear combinations of reward models, and then a shaped MLP can be defined in terms of these.

For AIRL you could then feed in an MLP reward model and an MLP potential shaping, it can internally sum them and have the train vs test distinction -- with no need for a separate class for this.

Note I've done something similar in https://github.com/HumanCompatibleAI/evaluating-rewards/blob/master/src/evaluating_rewards/rewards/base.py which has been working well (and I iterated on a couple of designs before getting to this stage). But that's all TF code, so cannot be used verbatim.

**Comment by ejnnr** [comment]
#311 mostly fixes these, except that IRL still uses its own reward models. Should be very easy to change so that IRL uses the models in `rewards.reward_nets`. But I don't see a training script for IRL. Do people just use `mce_irl` directly or is there some other place we'd need to modify?

**Comment by AdamGleave** [comment]
I don't think we have a training script for `tabular_irl`. Which is bad, we should, but I think that can probably be postponed -- it'd be a natural thing to do if/when we unify training different algorithms into a single script. So I think the only other place to update is `test_tabular`.
